pyvkfft/opencl.py

```diff
@@ -155,20 +155,60 @@ def _make_config(self):
                                          n_batch, skipx, skipy, skipz)
 
-    def fft(self, src: cla.Array, dest: cla.Array = None):
+    def fft(self, src: cla.Array, dest: cla.Array = None, queue: cl.CommandQueue = None):
         """
         Compute the forward FFT
         :param src: the source pyopencl Array
         :param dest: the destination pyopencl Array. Should be None for an inplace transform
+        :param queue: the pyopencl CommandQueue to use for the transform. If not given,
+            the queue given when constructing the application is used.
         :raises RuntimeError: in case of a GPU kernel launch error
         :return: the transformed array. For a R2C inplace transform, the complex view of the
             array is returned.
         """
+        if not queue:
+            if dest is None:
+                if src.queue is None:
+                    warnings.warn("queue is not given and the source array does not have a queue
+                        attached to it. Falling back to the queue given to the constructor
+                        of VkFFTApp. This is deprecated and will stop working in the future.
+                        Use src_array.with_queue(queue) to attach a queue to the array or
+                        pass a queue to this method", DeprecationWarning)
+                    queue = self.queue
+                else:
+                    queue = src.queue
+            elif dest.queue is None and src.queue is None:
+                warnings.warn("queue is not given and the source/dest arrays do not have a queue
+                        attached to it. Falling back to the queue given to the constructor
+                        of VkFFTApp. This is deprecated and will stop working in the future.
+                        Use src_array.with_queue(queue) to attach a queue to the array or
+                        pass a queue to this method", DeprecationWarning)
+                queue = self.queue
+            elif dest.queue != src.queue:
+                if dest.queue is None:
+                    queue = src.queue
+                elif src.queue is None:
+                    queue = dest.queue
+                else:
+                    warnings.warn("queue is not given and the source/dest arrays are not equal.
+                        Falling back to the queue given to the constructor. This is deprecated
+                        and will stop working in the future.
+                        Use src_array.with_queue(queue) to attach a queue to the array or
+                        pass a queue to this method", DeprecationWarning)
+                    queue = self.queue
+            else:
+                queue = src.queue
+
+            if queue != self.queue:
+                # TODO: warn here if self.queue has not been finished yet.
+                # There is no way to check that a queue is finished in pyopencl.
+                pass
```

Comment:
@inducer, is there a way to check if a queue has finished?

Reply:
If you have the events, you can query their execution status, but I'm not sure that that comes with specific coherency guarantees. I think only finishing the queue will provide the guarantees you might be looking for here.

Reply:
If I call finish on `self.queue`, I run into the error `CommandQueueUsedAfterExit: Command queue used after exit of context manager. This is deprecated and will stop working in 2023.`

Reply:
I'm confused because that should be a warning, not an error.

But yes, in essence, what that's saying is that you should make up your mind: either do not use the queue as a context manager (and thereby acknowledge that it is long-lived), or stop using it after the context manager exits. Forcing you to make that decision is the intention of the warning.

Reply:
Fixed with latest pyopencl